File: src/main/java/thedarkcolour/futuremc/block/BeeHiveBlock.java
package thedarkcolour.futuremc.block;

import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.SoundType;
import net.minecraft.block.material.Material;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.enchantment.Enchantments;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.fluid.IFluidState;
import net.minecraft.item.BlockItemUseContext;
import net.minecraft.state.StateContainer;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.Direction;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockReader;
import net.minecraft.world.World;
import net.minecraftforge.common.ToolType;
import thedarkcolour.futuremc.entity.BeeEntity;
import thedarkcolour.futuremc.tile.BeeHiveTileEntity;
import thedarkcolour.futuremc.tile.Tile;

import java.util.List;

import static thedarkcolour.futuremc.block.Props.FACING;
import static thedarkcolour.futuremc.block.Props.HONEY_LEVEL;

public class BeeHiveBlock extends InteractionBlock {
    public BeeHiveBlock(boolean isNest) {
        super(Block.Properties.create(Material.WOOD).sound(SoundType.WOOD).hardnessAndResistance(isNest ? 0.3F : 0.6F));
        setRegistryName(isNest ? "bee_nest" : "beehive");
        setDefaultState(getDefaultState().with(FACING, Direction.NORTH).with(HONEY_LEVEL, 0));
    }

    @Override
    public ToolType getHarvestTool(BlockState state) {
        return ToolType.AXE;
    }

    @Override
    public TileEntity createTileEntity(BlockState state, IBlockReader world) {
        return Tile.BEEHIVE.create();
    }

    @Override
    protected void fillStateContainer(StateContainer.Builder<Block, BlockState> builder) {
        builder.add(FACING, HONEY_LEVEL);
    }

    @Override
    public boolean hasComparatorInputOverride(BlockState state) {
        return true;
    }

    @Override
    public int getComparatorInputOverride(BlockState blockState, World worldIn, BlockPos pos) {
        if (worldIn.getTileEntity(pos) instanceof BeeHiveTileEntity) {
            return worldIn.getBlockState(pos).get(HONEY_LEVEL);
        }

        return 0;
    }

    @Override
    public boolean removedByPlayer(BlockState state, World worldIn, BlockPos pos, PlayerEntity player, boolean willHarvest, IFluidState fluid) {
        if (!worldIn.isRemote) {
            TileEntity tile = worldIn.getTileEntity(pos);
            if (tile instanceof BeeHiveTileEntity && EnchantmentHelper.getEnchantmentLevel(Enchantments.SILK_TOUCH, player.getHeldItemMainhand()) == 0) {
                ((BeeHiveTileEntity)tile).angerBees(player, BeeHiveTileEntity.BeeState.DELIVERED);
                worldIn.updateComparatorOutputLevel(pos, this);
            }

            List<BeeEntity> bees = worldIn.getEntitiesWithinAABB(BeeEntity.class, (new AxisAlignedBB(pos)).expand(8.0D, 6.0D, 8.0D));
            if (!bees.isEmpty()) {
                List<PlayerEntity> players = worldIn.getEntitiesWithinAABB(PlayerEntity.class, (new AxisAlignedBB(pos)).expand(8.0D, 6.0D, 8.0D));
                int size = players.size();

                for (BeeEntity bee : bees) {
                    if (bee.getAttackTarget() == null) {
                        bee.setBeeAttacker(players.get(worldIn.rand.nextInt(size)));
                    }
                }
            }
        }

        return super.removedByPlayer(state, worldIn, pos, player, willHarvest, fluid);
    }

    @Override
    public BlockState getStateForPlacement(BlockItemUseContext context) {
        return getDefaultState().with(FACING, context.getPlacementHorizontalFacing().getOpposite());
    }
}

File: src/main/java/thedarkcolour/futuremc/block/buzzybees/BeeHiveBlock.kt
package thedarkcolour.futuremc.block.buzzybees

import net.minecraft.block.BlockHorizontal
import net.minecraft.block.properties.PropertyBool
import net.minecraft.block.properties.PropertyDirection
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Blocks
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import thedarkcolour.core.block.InteractionBlock
import thedarkcolour.futuremc.registry.FBlocks.BEEHIVE
import thedarkcolour.futuremc.tile.BeeHiveTile
import java.util.*

class BeeHiveBlock(properties: Properties) : InteractionBlock(properties) {
    init {
        defaultState = defaultState.withProperty(FACING, EnumFacing.NORTH).withProperty(IS_FULL, false)
    }

    override fun getItemDropped(state: IBlockState, rand: Random, fortune: Int): Item {
        return if (state.block == BEEHIVE) super.getItemDropped(
            state,
            rand,
            fortune
        ) else Item.getItemFromBlock(Blocks.AIR)
    }

    override fun canSilkHarvest(world: World, pos: BlockPos, state: IBlockState, player: EntityPlayer): Boolean {
        return true
    }

    override fun getSilkTouchDrop(state: IBlockState): ItemStack {
        return ItemStack.EMPTY
    }

    override fun getHarvestTool(state: IBlockState): String {
        return "axe"
    }

    override fun createTileEntity(worldIn: World?, state: IBlockState?): TileEntity {
        return BeeHiveTile()
    }

    override fun createBlockState(): BlockStateContainer {
        return BlockStateContainer(this, FACING, IS_FULL)
    }

    override fun getStateFromMeta(meta: Int): IBlockState {
        return defaultState.withProperty(IS_FULL, meta and 4 != 0)
            .withProperty(FACING, EnumFacing.byHorizontalIndex(meta and -5))
    }

    override fun getMetaFromState(state: IBlockState): Int {
        return (if (state.getValue(IS_FULL)) 4 else 0) or state.getValue(FACING).horizontalIndex
    }

    override fun hasComparatorInputOverride(state: IBlockState): Boolean {
        return true
    }

    override fun getComparatorInputOverride(blockState: IBlockState, worldIn: World, pos: BlockPos): Int {
        return if (worldIn.getTileEntity(pos) is BeeHiveTile) {
            (worldIn.getTileEntity(pos) as BeeHiveTile).honeyLevel
        } else 0
    }

    override fun getStateForPlacement(
        worldIn: World,
        pos: BlockPos,
        facing: EnumFacing,
        hitX: Float,
        hitY: Float,
        hitZ: Float,
        meta: Int,
        placer: EntityLivingBase
    ): IBlockState {
        return defaultState.withProperty(FACING, placer.horizontalFacing.opposite)
    }

    override fun setHardness(hardness: Float): BeeHiveBlock {
        super.setHardness(hardness)
        return this
    }

    companion object {
        val IS_FULL = PropertyBool.create("full")
        @kotlin.jvm.JvmField
        val FACING: PropertyDirection = BlockHorizontal.FACING
    }
}